plugins/services/src/js/components/ConfigurationMapKeyValueRow.js

```diff
@@ -0,0 +1,115 @@
+import React from 'react';
+
+import ConfigurationMapLabel from '../../../../../src/js/components/ConfigurationMapLabel';
+import ConfigurationMapRow from '../../../../../src/js/components/ConfigurationMapRow';
+import ConfigurationMapValue from '../../../../../src/js/components/ConfigurationMapValue';
+
+class ConfigurationMapKeyValueRow extends React.Component {
+  render() {
+    let {autoHide, label, map, placeholder, type, value} = this.props;
+
+    // When a function is given, evaulate it at render time
+    if (typeof label === 'function') {
+      label = label();
+    }
+    if (typeof value === 'function') {
+      value = value();
+    }
+
+    // When the value is missing, decide how to handle it
+    if (value == null) {
+      if (autoHide) {
+        return null;
+      }
+
+      // If we want to keep the value, replace it with a long dash
+      value = placeholder;
+
+    } else {
```

Comment:
This is a huge if-else statement with switches and nested if-elses. Let's split it up